wrappers/desktop.go

```diff
@@ -102,32 +102,72 @@ var isValidDesktopFileLine = regexp.MustCompile(strings.Join([]string{
 	"^TargetEnvironment=",
 }, "|")).Match
 
-// rewriteExecLine rewrites a "Exec=" line to use the wrapper path for snap application.
-func rewriteExecLine(s *snap.Info, desktopFile, line string) (string, error) {
+type execRewriterImpl struct {
+	// strict enables fallback to matching a desktop file with snap
+	// application using the application's name
+	strict bool
+	// rewriter matches given app with a desktop file and returns a rewritten
+	// exec line or an empty string
+	rewriter func(app *snap.AppInfo, desktopFile, execLine string) string
+}
+
+// appExecRewriter matches a snap app with desktop file and an exec command by
+// looking at snap command wrapper, returns an updated command or an empty
+// string
+func appExecRewriter(app *snap.AppInfo, desktopFile, execCmd string) string {
 	env := fmt.Sprintf("env BAMF_DESKTOP_FILE_HINT=%s ", desktopFile)
 
+	wrapper := app.WrapperPath()
+	validCmd := filepath.Base(wrapper)
+	// check the prefix to allow %flag style args
+	// this is ok because desktop files are not run through sh
+	// so we don't have to worry about the arguments too much
+	if execCmd == validCmd {
+		return env + wrapper
+	} else if strings.HasPrefix(execCmd, validCmd+" ") {
+		return fmt.Sprintf("%s%s%s", env, wrapper, execCmd[len(validCmd):])
+	}
+	return ""
+}
+
+// appAutostartExecRewriter matches a snap app with a desktop file and an exec
+// command by inspecting the application's autostart declaration, returns an
+// updated command or an empty string
+func appAutostartExecRewriter(app *snap.AppInfo, desktopFile, execCmd string) string {
+	if app.Autostart != filepath.Base(desktopFile) {
+		return ""
+	}
+
+	cmd := execCmd
+	if pos := strings.Index(execCmd, " "); pos != -1 {
```

Comment:
Why is this even different from the case above?  It sounds like HasPrefix above aims at catching the same case, doesn't it?